packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -3732,7 +3776,16 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
     bool seenTopActiveRoute = false; // Whether we've seen the route that would get didPopNext.
     final List<_RouteEntry> toBeDisposed = <_RouteEntry>[];
     while (index >= 0) {
-      switch (entry!.currentState) {
+      if (entry!.pendingDecision != null) {
+        entry.flushPendingDecision();
+        assert(entry.pendingDecision == null && entry.currentState != _RouteLifecycle.staging);
+        if (entry.currentState == _RouteLifecycle.disposed){
+          // The route is exited synchronously, and the history has finished
+          // updated along with it. No need to walk the list again.
```

Comment:
updated -> updating

Reply:
It's not clear to me why we can abort flushing the history when one route is disposed. Why do the other routes not need to be processed?

Reply:
If the pop finished synchronously, the _flushHistoryUpdate will be called again at the end of it. that means the rest of the stack would have been updated.